memgpt/memory.py

```diff
@@ -16,96 +18,214 @@
+        super().__setattr__(name, value)
+        if name == "value":
+            # run validation
+            self.__class__.validate(self.dict(exclude_unset=True))
 
-    def __init__(self, persona=None, human=None, persona_char_limit=None, human_char_limit=None, archival_memory_exists=True):
-        self.persona = persona
-        self.human = human
-        self.persona_char_limit = persona_char_limit
-        self.human_char_limit = human_char_limit
+    @validator("value", always=True)
+    def check_value_length(cls, v, values):
+        if v is not None:
+            # Fetching the limit from the values dictionary
+            limit = values.get("limit", 2000)  # Default to 2000 if limit is not yet set
 
-        # affects the error message the AI will see on overflow inserts
-        self.archival_memory_exists = archival_memory_exists
+            # Check if the value exceeds the limit
+            if isinstance(v, str):
+                length = len(v)
+            elif isinstance(v, list):
+                length = sum(len(item) for item in v)
+            else:
+                raise ValueError("Value must be either a string or a list of strings.")
 
-    def __repr__(self) -> str:
-        return f"\n### CORE MEMORY ###" + f"\n=== Persona ===\n{self.persona}" + f"\n\n=== Human ===\n{self.human}"
+            if length > limit:
+                error_msg = f"Edit failed: Exceeds {limit} character limit (requested {length})."
+                # TODO: add archival memory error?
+                raise ValueError(error_msg)
+        return v
+
+    def __len__(self):
+        return len(str(self))
+
+    def __str__(self) -> str:
+        if isinstance(self.value, list):
+            return ",".join(self.value)
+        elif isinstance(self.value, str):
+            return self.value
+        else:
+            return ""
+
+
+class BaseMemory:
+
+    def __init__(self):
+        self.memory = {}
+
+    @classmethod
+    def load(cls, state: dict):
+        """Load memory from dictionary object"""
+        obj = cls()
+        for key, value in state.items():
+            obj.memory[key] = MemoryModule(**value)
+        return obj
+
+    def __str__(self) -> str:
+        """Representation of the memory in-context"""
+        section_strs = []
+        for section, module in self.memory.items():
+            section_strs.append(f'<{section} characters="{len(module)}/{module.limit}">\n{module.value}\n</{section}>')
+        return "\n".join(section_strs)
 
     def to_dict(self):
-        return {
-            "persona": self.persona,
-            "human": self.human,
+        """Convert to dictionary representation"""
+        return {key: value.dict() for key, value in self.memory.items()}
+
+
+class ChatMemory(BaseMemory):
+
+    def __init__(self, persona: str, human: str, limit: int = 2000):
+        self.memory = {
+            "persona": MemoryModule(name="persona", value=persona, limit=limit),

```

Comment:
Ahh ok, so is a `MemoryModule` the contents of a section within a `BaseMemory`? 

If I can make a suggestion, the BaseMemory.memory structure is awkward as a dict of typed objects with duplicate props. It seems like it should be a list (since that also preserves order, in case you care about that). 

Something like 

```python
>> ChatMemory.thoughts
[<class 'MemoryModule' persona 145>,
 <class 'MemoryModule' human 492>,
 <class 'MemoryModule' company 941>]
``` 
There's a data normalization issue with the duplication here, you can have this:

```python
>> chat_memory.memory.persona
<class 'MemoryModule' human 1998> # This shouldn't be possible!
```

If it's about lookups, there's some easy metaprogramming we can do for that 

```python

def __getattr__(self, val):
    try:
        return next([m in self.thoughts if m.name == 'val'])
    except StopIteration as e:
        raise AttributeError(f'{self.__class__.__name__} has no attribute {val}') from e
```
which looks like this in practice

```python
>> chat_memory.persona
<class 'MemoryModule' persona 2401>

>> chat_memory.three_laws
<class 'MemoryModule' three_laws 100>

>> chat_memory.hamburger
AttributeError: ChatMemory has no attribute 'hamburger'
```

there would need to be a little extra to enforce only a single MemoryModule of a given type, unless you want to be able to have more than one persona or human? worth a thought

Reply:
Yeah the `MemoryModule` basically represents a reserved section of the context (or at least allocation of a context up to the size `limit`) for placing the actual text value of the section of memory - we will also probably generalize this eventually to also include the system prompt (not just core memory), and store things like pre-defined human/persona values or system prompts under the same scheme. 

We also considered `Variable`, `Snipped`, and `ContextSlice` - `Thoughts` overlaps with inner-thoughts which is very different so I think that might be confusing.

Reply:
^^^ bingo! if we can add this into the code even as a docstring, it'll help when the name is a little sticky. 

So if a `Memory` is the entire context window (like the `core memory` computing paradigm), 
are these `slots` or `registers` or `MemoryPartials`... 


But I have held this PR up enough, and perfect is the enemy of gsd. I vote roll with `MemoryModule` until a more natural object name comes to mind, then migrate.